lib/stm32/common/fdcan.c

```diff
@@ -0,0 +1,763 @@
+ */
+
+#include <libopencm3/stm32/fdcan.h>
+#include <libopencm3/stm32/rcc.h>
+
+
+/* --- FD-CAN internal functions -------------------------------------------- */
+
+/** Routine implementing FDCAN_CCCR's INIT bit manipulation.
+ * This routine will change INIT bit and wait for it to actually
+ * change its value. If change won't happen before timeout,
+ * error is signalized. If INIT bit already has value which
+ * should be set, this function will return immediately.
+ *
+ * @param [in] canport FDCAN block base address
+ * @param [in] set new value of INIT, true means set
+ * @returns FDCAN_E_OK on success, FDCAN_E_TIMEOUT if INIT bit value
+ * didn't change before timeout has expired.
+ */
+static int fdcan_cccr_init_cfg(uint32_t canport, bool set, uint32_t timeout)
+{
+	uint32_t expected;
+	uint32_t wait_ack;
+
+	if (set) {
+		if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == FDCAN_CCCR_INIT) {
+			/* Already there, sir */
+			return FDCAN_E_OK;
+		}
+
+		FDCAN_CCCR(canport) |= FDCAN_CCCR_INIT;
+		expected = FDCAN_CCCR_INIT;
+	} else {
+		if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == 0) {
+			/* Already there, sir */
+			return FDCAN_E_OK;
+		}
+
+		FDCAN_CCCR(canport) &= ~FDCAN_CCCR_INIT;
+		expected = 0;
+	}
+
+	/* Wait, until INIT bit is acknowledged */
+	wait_ack = timeout;
+	while ((--wait_ack) &&
+			((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == expected));
+
+	if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == expected) {
+		return FDCAN_E_OK;
+	} else {
+		return FDCAN_E_TIMEOUT;
+	}
+}
+
+/** Return ID of next free Tx buffer.
+ * Examines transmit buffer allocation in message RAM
+ * and returns ID of buffer, which is free.
+ *
+ * @param [in] canport FDCAN block base address
+ * @returns Non-negative number ID of Tx buffer which is free,
+ * or FDCAN_E_BUSY if no Tx buffer is available
+ */
+static int fdcan_get_free_txbuf(uint32_t canport)
+{
+	if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP0) == 0) {
+		return 0;
+	} else if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP1) == 0) {
+		return 1;
+	} else if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP2) == 0) {
+		return 2;
+	}
+
+	return FDCAN_E_BUSY;
+}
+
+/** Returns fill state and next available get index from receive FIFO.
+ * Examines FDCAN receive FIFO and returns fill status of FIFO and ID of
+ * next message available for reading. If fill status is 0 (FIFO is empty),
+ * then get index is undefined.
+ *
+ * @param [in] canport FDCAN block base address
+ * @param [in] fifo_id ID of fifo queried (currently 0 or 1)
+ * @param [out] get_index Address of buffer where next get index will be stored
+ * @param [out] pending_frames Address of buffer where amount of pending frame will be stored.
+ * @returns Operation error status.
+ */
+static int fdcan_get_fill_rxfifo(uint32_t canport, uint8_t fifo_id, unsigned *get_index,
+		unsigned *pending_frames)
+{
+	if (fifo_id == 0) {
+		*get_index = (FDCAN_RXF0S(canport) & FDCAN_RXF0S_F0GI_MASK)
+			>> FDCAN_RXF0S_F0GI_SHIFT;
```

Comment:
eg, using unshifted masks, all that changes is that you shift before hand.
(leaving review here, have to siwtch tasks)